src/satosa/micro_services/custom_logging.py

```diff
@@ -73,12 +74,13 @@ def process(self, context, data):
             # This is where the logging magic happens
             log = {}
             log['router'] = context.state.state_dict['ROUTER']
-            log['timestamp'] = data.auth_info.timestamp
+            log['samltime'] = data.auth_info.timestamp
+            log['timestamp'] = datetime.now().isoformat()
```

Comment:
`datetime.now().isoformat()` will give you the localtime that the server has been configured with. I would advise against that. The only time that the system should work with is UTC. Anything other than UTC is for interfacing. I suggest this to be

```suggestion
            log['timestamp'] = datetime.utcnow().isoformat()
```

Reply:
I do not disagree with this change, but do check if this will not yield 'double' data in the statistics as upon deploying this in prod, as the log data will travel 'back in time'. The logs parser does not know about timezones etc so it will just add the data at the given date. Not a problem if that is noon, but possibly not nice f that is midnight

Reply:
I was thinking the same: we actually really _want_ the local server time, as that is the time that the transaction took place and what our logparser uses to store the transaction in the database. If we were to use UTC here, we would need to adjust the log parser as well, and that would then need to be able to distinguish old logs from new logs during the 28 day log preserving window?

Reply:
https://github.com/inacademia-development/audit/blob/76e78497b5e5daab9e207226a91f85e2dd4bf533/audit_log_handler_docker/app/parse_audit_logs.py#L43

Reply:
SAML messages encode time in UTC. The timestamp that was previously saved (`data.auth_info.timestamp`) should already be in UTC (unless other processing takes place that converts it to local time.)

> we actually really want the local server time

What we need is a stable timeline. This is what UTC provides. Local time is a projection off of UTC; it is always a representation based on external rules.
Depending on local time means that you depend on the host environment having been configured the "right" way; something that the application cannot inspect, or guarantee. This creates an implicit dependency that can turn into a big mess if somebody changes the clock settings.
Additionally, you depend on those external rules that dictate how this timezone representation changes in relation to UTC; you depend on the projection parameters. Those rules/parameters are non-technical and unstable; they are based on political decisions and then encoded in a global database which must be kept up-to-date, and in sync across all hosts that work with local time.

For those reasons, the only time that a system should know is UTC. Anything else is a configurable projection from that stable timeline. Timezones are representations of UTC. All processing and computation is done in UTC. Interfaces can be configured to show the local time, but the system only deals with time in UTC.

Reply:
Let it be clear that I never doubted the merits of UTC timestamps. The problem is what the effects downstream may be. parse_audit _logs.py does this with the timestamp it finds in the logs:
```
def parse_timezone(datetime):
    t = dateutil.parser.parse(datetime)
    return t.strftime('%Y-%m-%d %H:%M:%S')
```
The return value is then used to store the log time in a mysql _datetime_ column. This string does not have any timezone reference and is probably pure UTC in the case where we feed it UTC unless parse(datetime) adds timezone info to UTC strings? This results in a _skewed_ impression of the statistics (in relation to the local server time) if the results are not compensated for the timezone adjustments in the stats portal, which is not the case when I look at this line in the stats portal:
```
function get_idpsessions($s, $e, $p, $f="", $sp="", $idp="") {
    ...
    $query .= "where l.log_timestamp between '$s' and '$e' ";
```  
Also: we have complete control over the server configuration, so timezone should never be wrong.
So, yes I agree we should have used UTC timestamps from the start, but given the downstream handling I think we _have to_ use now() untill we fix the complete train?

Or, you see it differently and I'd like to be enlightened...

Reply:
Hmm... when testing in my dev env I have an IdP that's configured for UTC on the console and says it's date is ```Wed Oct 21 07:34:16 UTC 2020```, it has however sent a SAML Response for ```2020-10-21T06:43:47Z``` a bit earlier and that's an hour off, unless Zulu corrects for that difference, which it doesn't according to parser.parse()?

Reply:
I can now at least confirm that the original ```data.auth_info.timestamp``` log contains the SAML ```AuthnInstant``` and not the ```IssueInstant``` which might explain the weird logtimes we have seen for which this PR is the intended solution.

Reply:
> so we have allways used UTC in datestamp
>  we have to decide if we want to continue doing so

right, so we should continue doing what we were doing. It was and is the right thing.

> the way the stats portal queries the DB, [which actually] is a totally different discussion

I agree. What the portal is doing with that data is a different discussion. If the portal needs to show the local-time then it should know what local-time is (the timezone should be configurable) and convert from UTC to that, and display it (not store it).

Whether this is an actual need or requirement we should discuss with the team.